src/genetic_low_memory.c

```diff
@@ -0,0 +1,459 @@
+  const unsigned int mergedParentsLength =
+      (dimensions * UINT16_DIGIT_COUNT) + 1;
+  while (currentNextGenerationSize < (populationSize)) {
+
+    unsigned int parent1Index, parent2Index = 0;
+    tourney(population, &parent1Index, &parent2Index, evaluationFunction);
+
+    uint16_t parent1[dimensions];
+    uint16_t parent2[dimensions];
+
+    for (int i = 0; i < dimensions; i++) {
+
+      parent1[i] = population[(parent1Index * dimensions) + i];
+      parent2[i] = population[(parent2Index * dimensions) + i];
+    }
+
+    char mergedParents1[mergedParentsLength];
+    encode(mergedParents1, parent1);
+    char mergedParents2[mergedParentsLength];
+    encode(mergedParents2, parent2);
+
+    createChildren(mergedParents1, mergedParents2, nextGeneration,
+                   &currentNextGenerationSize);
+  }
+}
+
+/**
+ * This algorithm uses the elitism principal and this function helps us keep
+ * track of the current most optimal solution
+ *
+ *
+ * @param population this array stores all the values of the population
+ * @param bestValuesCoordinates this array stores the parameters of the current
+ * best solution
+ * @param index  this represents the index of the new best solution
+ * @param storedValue this represents the current most optimal fitness
+ * @param newValue this represents the new superior fitness
+ */
+static void replaceEliteValue(uint16_t* population,
+                              uint16_t* bestValuesCoordinates,
+                              unsigned int index, float* storedValue,
+                              float newValue) {
+
+  *storedValue = newValue;
+  for (unsigned int i = 0; i < dimensions; i++) {
+    bestValuesCoordinates[i] = population[index + i];
+  }
+}
+
+/**
+ * This function calculates and stores the two best solutions of the population
+ *
+ * @param population this array stores the values of eaech parameter of the
+ * population
+ * @param bestFit this is the best fitness that was calculated
+ * @param bestFitCoord these are the parameters of the best solution
+ * @param evaluationFunction this function is used to evaluate each solution
+ * @param secondBestValues  this is the second best fitness that was calculated
+ * @param secondBestFitness  these are the parameters of the second solution
+ */
+static void calculateFitness(uint16_t* population, float* bestFit,
+                             uint16_t* bestFitCoord,
+                             float (*evaluationFunction)(float*),
+                             uint16_t* secondBestValues,
+                             float* secondBestFitness) {
+
+  for (unsigned int i = 0; i < populationSize; i++) {
+
+    float parameters[dimensions];
+    const unsigned int baseIndex = i * dimensions;
+
+    for (unsigned int j = 0; j < dimensions; j++) {
+      parameters[j] = ((float)population[baseIndex + j] * UINT16_MAX_INVERSE);
+    }
+
+    const float strength = evaluationFunction(parameters);
+
+    if (strength < *(bestFit)) {
+
```

Comment:
Why don't we update bestFit here? What am I missing ?

Reply:
The replaceEliteValue function takes care of updating it

Reply:
I don't think it should. This method should take care of updating the elitePool, not updating a random variable that you give it. It would be more efficient to do the assignment outside of the function. It would also reduce the number of arguments which would make the code easier to read.

Do you agree ?

Reply:
Thinking about it might be better to remove the whole function , because without the assignation it is simply a memcpy